gluetool/log.py

```diff
@@ -605,6 +624,72 @@ def _add_context(context_name, context_value):
         return msg
 
 
+class JSONLoggingFormatter(logging.Formatter):
+    """
+    Custom logging formatter producing a JSON dictionary describing the log record.
+    """
+
+    def __init__(self, **kwargs):
+        # pylint: disable=unused-argument
+
+        super(JSONLoggingFormatter, self).__init__()
+
+    def format(self, record):
+        # Construct a huuuuge dictionary describing the event
+        serialized = {
+            field: getattr(record, field, None)
+            for field in (
+                # LogRecord properties
+                'args', 'created', 'exc_info', 'exc_text', 'filename', 'funcName', 'levelname', 'levelno',
+                'lineno', 'module', 'msecs', 'msg', 'name', 'pathname', 'process', 'processName',
+                'relativeCreated', 'thread', 'threadName',
+                # our custom fields
+                'raw-blob', 'raw-struct', 'raw-xml'
+            )
+        }
+
+        serialized['message'] = record.getMessage()
+
+        if record.exc_info:
+            serialized['caused_by'] = []
+
+            exc_info = record.exc_info
+
+            # pylint: disable=invalid-name
+            def _add_cause(exc, tb):
+                exc_class = exc.__class__
+
+                stack = _extract_stack(tb)
+
+                serialized['caused_by'].append({
+                    'exception': {
+                        'class': '{}.{}'.format(exc_class.__module__, exc_class.__name__),
+                        'message': exc.message
+                    },
+                    'traceback': [
+                        {
+                            'filename': filename,
+                            'lineno': lineno,
+                            'fnname': fnname,
+                            'text': text,
+                            'locals': frame.f_locals
+                        }
+                        for filename, lineno, fnname, text, frame in stack
+                    ]
+                })
+
+            # don't unpack traceback - it might lead to a circular reference, leaving this frame
+            # uncollectable
```

Comment:
I don`t catch what this comment refers to

Reply:
and all the `exc_info[..]` stuff below is too much magical

Reply:
I added a few more comments to describe what's supposed to happen.